src/lib/utils/plugin_manager.cpp

```diff
@@ -0,0 +1,88 @@
+#include <dlfcn.h>
+
+#include "storage/storage_manager.hpp"
+#include "utils/abstract_plugin.hpp"
+#include "utils/assert.hpp"
+#include "utils/filesystem.hpp"
+
+#include "plugin_manager.hpp"
+
+namespace opossum {
+
+bool PluginManager::_is_duplicate(AbstractPlugin* plugin) const {
+  // This should work as soon as we support gcc-8 or gcc-8 supports us (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86740)
+  // for ([[maybe_unused]] auto &[plugin_name, plugin_handle_wrapper] : _plugins) {
+  //   if (plugin_handle_wrapper.plugin == plugin) {
+  //     return true;
+  //   }
+  // }
+
+  // return false;
+
+  for (const auto& p : _plugins) {
+    auto plugin_handle_wrapper = p.second;
+    if (plugin_handle_wrapper.plugin == plugin) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void PluginManager::load_plugin(const filesystem::path& path) {
+  const auto name = plugin_name_from_path(path);
+
+  Assert(!_plugins.count(name), "Loading plugin failed: A plugin with name  " + name + " already exists.");
+
+  PluginHandle plugin_handle = dlopen(path.c_str(), static_cast<uint8_t>(RTLD_NOW) | static_cast<uint8_t>(RTLD_LOCAL));
+  Assert(plugin_handle, "Loading plugin failed: " + dlerror());
+
+  // abstract_plugin.hpp defines a macro for exporting plugins which makes them instantiable by providing a
+  // factory method. See the sources of AbstractPlugin and TestPlugin for further details.
+  void* factory = dlsym(plugin_handle, "factory");
+  Assert(factory, "Instantiating plugin failed: Have you implemented and exported the factory method?");
+
+  using Instantiator = AbstractPlugin* (*)();
```

Comment:
I really don't know how to better describe that, any suggestion?

Reply:
Why not `using Factory`? But then, it is not really a factory, because it does not take any arguments and only returns a single object, ever.

`SingletonGetter` and `singleton_get()`?